app/code/Magento/Shipping/Controller/Adminhtml/Order/Shipment/AddTrack.php

```diff
@@ -18,56 +31,84 @@ class AddTrack extends \Magento\Backend\App\Action
     const ADMIN_RESOURCE = 'Magento_Sales::shipment';
 
     /**
-     * @var \Magento\Shipping\Controller\Adminhtml\Order\ShipmentLoader
+     * @var ShipmentLoader
      */
     protected $shipmentLoader;
 
+    /**
+     * @var ShipmentRepositoryInterface
+     */
+    private $shipmentRepository;
+
+    /**
+     * @var ShipmentTrackInterfaceFactory
+     */
+    private $trackFactory;
+
+    /**
+     * @var SerializerInterface
+     */
+    private $serializer;
+
     /**
      * @param Action\Context $context
-     * @param \Magento\Shipping\Controller\Adminhtml\Order\ShipmentLoader $shipmentLoader
+     * @param ShipmentLoader $shipmentLoader
+     * @param ShipmentRepositoryInterface|null $shipmentRepository
+     * @param ShipmentTrackInterfaceFactory|null $trackFactory
+     * @param SerializerInterface|null $serializer
      */
     public function __construct(
         Action\Context $context,
-        \Magento\Shipping\Controller\Adminhtml\Order\ShipmentLoader $shipmentLoader
+        ShipmentLoader $shipmentLoader,
+        ShipmentRepositoryInterface $shipmentRepository = null,
+        ShipmentTrackInterfaceFactory $trackFactory = null,
+        SerializerInterface $serializer = null
     ) {
-        $this->shipmentLoader = $shipmentLoader;
         parent::__construct($context);
+
+        $this->shipmentLoader = $shipmentLoader;
+        $this->shipmentRepository = $shipmentRepository ?: ObjectManager::getInstance()
+            ->get(ShipmentRepositoryInterface::class);
+        $this->trackFactory = $trackFactory ?: ObjectManager::getInstance()
+            ->get(ShipmentTrackInterfaceFactory::class);
+        $this->serializer = $serializer ?: ObjectManager::getInstance()
+            ->get(SerializerInterface::class);
     }
 
     /**
-     * Add new tracking number action
+     * Add new tracking number action.
      *
-     * @return void
-     * @throws \Magento\Framework\Exception\LocalizedException
+     * @return ResultInterface
      */
-    public function execute()
+    public function execute(): ResultInterface
```

Comment:
It's a good idea to add strict type to the method but unfortunately, we are not allowed to change public methods signature according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)